src/order/locally_finite.lean

```diff
@@ -147,13 +163,87 @@ def locally_finite_order.of_Icc (α : Type*) [partial_order α] [decidable_eq α
   finset_mem_Ioo := λ a b x, by rw [finset.mem_filter, mem_Icc, and_and_and_comm, lt_iff_le_and_ne,
     lt_iff_le_and_ne] }
 
+/-- A constructor from a definition of `finset.Iic` alone, the other ones being derived by removing
+the ends. As opposed to `locally_finite_order_top.of_Ici`, this one requires `decidable_rel (≤)` but
+only `preorder`. -/
+def locally_finite_order_top.of_Ici' (α : Type*) [preorder α] [decidable_rel ((≤) : α → α → Prop)]
+  (finset_Ici : α → finset α) (mem_Ici : ∀ a x, x ∈ finset_Ici a ↔ a ≤ x) :
+  locally_finite_order_top α :=
+{ finset_Ici := finset_Ici,
+  finset_Ioi := λ a, (finset_Ici a).filter (λ x, ¬x ≤ a),
+  finset_mem_Ici := mem_Ici,
+  finset_mem_Ioi := λ a x, by rw [mem_filter, mem_Ici, lt_iff_le_not_le] }
+
+/-- A constructor from a definition of `finset.Iic` alone, the other ones being derived by removing
+the ends. As opposed to `locally_finite_order_top.of_Ici'`, this one requires `partial_order` but
+only `decidable_eq`. -/
+def locally_finite_order_top.of_Ici (α : Type*) [partial_order α] [decidable_eq α]
+  (finset_Ici : α → finset α) (mem_Ici : ∀ a x, x ∈ finset_Ici a ↔ a ≤ x) :
+  locally_finite_order_top α :=
+{ finset_Ici := finset_Ici,
+  finset_Ioi := λ a, (finset_Ici a).filter (λ x, a ≠ x),
+  finset_mem_Ici := mem_Ici,
+  finset_mem_Ioi := λ a x, by rw [mem_filter, mem_Ici, lt_iff_le_and_ne] }
+
+/-- A constructor from a definition of `finset.Iic` alone, the other ones being derived by removing
+the ends. As opposed to `locally_finite_order.of_Icc`, this one requires `decidable_rel (≤)` but
+only `preorder`. -/
+def locally_finite_order_bot.of_Iic' (α : Type*) [preorder α] [decidable_rel ((≤) : α → α → Prop)]
+  (finset_Iic : α → finset α) (mem_Iic : ∀ a x, x ∈ finset_Iic a ↔ x ≤ a) :
+  locally_finite_order_bot α :=
+{ finset_Iic := finset_Iic,
+  finset_Iio := λ a, (finset_Iic a).filter (λ x, ¬a ≤ x),
+  finset_mem_Iic := mem_Iic,
+  finset_mem_Iio := λ a x, by rw [mem_filter, mem_Iic, lt_iff_le_not_le] }
+
+/-- A constructor from a definition of `finset.Iic` alone, the other ones being derived by removing
+the ends. As opposed to `locally_finite_order_top.of_Ici'`, this one requires `partial_order` but
+only `decidable_eq`. -/
+def locally_finite_order_top.of_Iic (α : Type*) [partial_order α] [decidable_eq α]
+  (finset_Iic : α → finset α) (mem_Iic : ∀ a x, x ∈ finset_Iic a ↔ x ≤ a) :
+  locally_finite_order_bot α :=
+{ finset_Iic := finset_Iic,
+  finset_Iio := λ a, (finset_Iic a).filter (λ x, x ≠ a),
+  finset_mem_Iic := mem_Iic,
+  finset_mem_Iio := λ a x, by rw [mem_filter, mem_Iic, lt_iff_le_and_ne] }
+
 variables {α β : Type*}
 
+/-- An empty type is locally finite. -/
+protected def _root_.is_empty.to_locally_finite_order [preorder α] [is_empty α] :
```

Comment:
```suggestion
/-- An empty type is locally finite. This is not an instance as it would form diamonds on
`locally_finite_order (α × β)` with `prod.locally_finite_order` when `is_empty α`.

See note [reducible non-instances]. -/
@[reducible] protected def _root_.is_empty.to_locally_finite_order [preorder α] [is_empty α] :
```

Reply:
And same for the two instances below